Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -240,51 +354,63 @@ public struct HTTPHeaders: CustomStringConvertible {
     /// - Parameter name: The header field name whose values are to be retrieved.
     /// - Returns: A list of the values for that header field name.
     public subscript(name: String) -> [String] {
-        if let result = storage[name.lowercased()] {
-            return result.map { tuple in tuple.1 }
+        guard !self.headers.isEmpty else {
+            return []
         }
-        return []
-    }
 
-    /// Serializes this HTTP header block to bytes suitable for writing to the wire.
-    ///
-    /// - Parameter buffer: A buffer to write the serialized bytes into. Will increment
-    ///     the writer index of this buffer.
-    func write(buffer: inout ByteBuffer) {
-        for (key, values) in storage {
-            if key != "set-cookie" {
-                writeListHeaderValues(buffer: &buffer, key: key, values: values)
-            } else {
-                writeSequentialHeaderValues(buffer: &buffer, key: key, values: values)
+        let utf8 = name.utf8
+        var array: [String] = []
+        for header in headers {
+            if self.buffer.equalCaseInsensitiveASCII(view: utf8, at: header.name) {
+                array.append(string(idx: header.value))
             }
         }
-        buffer.write(staticString: crlf)
+        return array
     }
 
-    /// Used for most HTTP headers, which can be represented as a single line joined by commas.
-    private func writeListHeaderValues(buffer: inout ByteBuffer, key: String, values: [(String, String)]) {
-        buffer.write(string: key)
-        buffer.write(staticString: headerSeparator)
+    /// Checks if a header is present
+    ///
+    /// - parameters:
+    ///     - name: The name of the header
+    ///     - value: The value of the header or nil if the name should not be checked
+    //  - returns: `true` if a header with the name (and value) exists, `false` otherwise.
+    public func contains(name: String, value: String? = nil) -> Bool {
+        guard !self.headers.isEmpty else {
+            return false
+        }
 
-        var writerIndex = buffer.writerIndex
-        for (_, value) in values {
-            buffer.write(string: value)
-            writerIndex = buffer.writerIndex
-            buffer.write(staticString: ",")
+        let utf8 = name.utf8
+        for header in headers {
+            if self.buffer.equalCaseInsensitiveASCII(view: utf8, at: header.name) {
+                if let value = value {
+                    return self.buffer.equalCase(string: value, at: header.name)
```

Comment:
This isn't right: it exits early. That means if you have a header block that consists of `[('hello', 'world'), ('hello', 'fish')]` and you ask `headers.contains(name: 'hello', value: 'fish')` this will return `false` instead of `true`. Mind adding a test that covers this case, and then fixing this function?